src/util/diproche/diprocheInterface.ts

```diff
@@ -0,0 +1,97 @@
+import Issue from "../../issueHandling/issue";
+import { concatOneIssueList } from "../../issueHandling/issueMapping";
+import { getAllIssues } from "../../vocabularyChecker/detectWrongSyntax";
+import { UnexpectedError } from "./Errors";
+import modes from "./predicateList.json";
+
+export type Mode = "propositionalLogic" | "firstOrderPredicateLogic" | "test";
+
+/**
+ * Since the user input most likely lacks the correct predicate,
+ * it needs to be added. For easier maintainability in case
+ * of an expanding predicate list, the corresponding predicate is
+ * added with respect to the current @param mode. This can, for
+ * example correspond to the concept of "Spielwiese". The user should
+ * be able to just stay inside a mode and not have to add this mode
+ * himself. Hence, the mode needs to be read out from the browser.
+ */
+export function addPredicate(userInput: string, mode: Mode) {
+	for (const predicate of modes.predicates) {
+		if (predicate === mode) {
+			return getPrologPredicateForMode(mode).concat("(").concat(userInput).concat(").");
```

Comment:
Do you mean it is less readable because concat needs parantheses itself? Otherwise I don't see how concat makes it less readable. In fact, the intention behind using concat was better readability, so one can immediately understand, that two strings are added.

Is your argument behind + being more readable, that there is less text and no additional parantheses which could be confusing?

Either way I think this is mostly a choice of preference, so I don't really care if I change it to + or leave it at concat.

Reply:
I meant it is less readable because of the parentheses.
I think using + to concate strings will not make it less readable to other developers though since this is pretty usual to do in javascript.

The only case were + should definitly be chosen over concat is when you call will call it many times for example in a loop, since + is faster than concat.

See here:
http://jsben.ch/D4itG

If you wanna address my comment and keep concat you could do something like this:
```
return getPrologPredicateForMode(mode)
  .concat("(")
  .concat(userInput)
  .concat(").");
```

Reply:
Interestingly enough I get a 2% faster result for concat when following your link, meaning concat is faster by the slightest bit. After running the tests a second time it was even 5% faster and after the 4th time running it, + was 5% faster. The variance seems to be even stronger, since after doing more test runs one was 20% faster than the other and then they swapped roles again.

The set of words on the left didn't seem to change. So I guess the results don't speak in favor of any one of those methods. Also, even one was faster, I don't think that's really the important thing, since I only use these methods twice, whereas the benchmark uses round about 500 words, thus using both methods around 500 times.

 However, since none of those methods seems to be faster and I most likely have author-bias, I will change it to +.

Reply:
Benchmarking JavaScript is terribly complicated; trust me, I interned in the V8-team for 3 months, and if there's anything I learned, it's that you've got to take into account way to many things. We tested and benchmarked on countless different platforms and with different optimization strategies, and there was rarely something that was "obviously better in every possible way".
Therefore, I would put emphasis on readability. Your code doesn't seem to be performance sensitive anyways, so that should really be a no-brainer. Simplest solution (with a very reasonable performance, potentially even faster than `concat` or `+` (for number of parameters known at compile time):
```typescript
`${getPrologPredicateForMode(mode)}(${userInput})`
```
Here the compiler and optimizer can (at compile time) see how many things get combined and can potentially choose the most efficient approach right away. It is both short and pretty readable, too.